### PR TITLE
Create alternative programming examples that use `dma_task` method for runtime sequence operations

### PR DESCRIPTION
**Pull request by hunhoffe**
This PR is an incremental step towards implementation of my IRON augmentations (found here: https://github.com/Xilinx/mlir-aie/pull/1732).

The IRON extensions/augmentations generate runtime sequences using dma tasks. However, I noticed that the dma task infrastructure does not seem to work out of the box for all examples. Thus, enabling the programming examples to use dma tasks is a necessary preliminary step before I can merge my other work.

While having alternative versions of the program examples will slow down the CI, I believe this is a temporary step until we are ready to deprecate examples using the current IRON syntax. I plan to replace the alt versions with jupyter notebooks eventually.

### CODE REVIEW

**Comment by hunhoffe** [comment]
I am seeing some failures with the programming examples when using the dma task interface. There are several separate errors I suspect so far:
- [ ] Verification: Does not accept single element sizes, e.g., `[1, 1, 1, 1]`, which is used in some examples of transfer a scalar value.
- [ ] Special Case: Does not accept sizes of form `[1, 1, 1, N]` where N >= 1024; this is a special case for `npu_dma_memcpy_nd` so hopefully not too hard to port
- [ ] Functionality: It looks like all examples with a repeat count/step in the upper dimension fail; there appears to be a `repeat_count` parameter in the call to `dma_configure_task_for` which I believe works; however, this is not sufficient for examples that also have an `iteration_step` value in the upper dimension of the strides.

**Comment by andrej** [comment]
Nice, I'm happy to see this feature getting some attention. Apologies in advance if you run into some roadblocks ... I believe this is the first time somebody besides me is seriously using this syntax., so it is unfortunately likely that some troubleshooting may be necessary.

>  Verification: Does not accept single element sizes, e.g., [1, 1, 1, 1], which is used in some examples of transfer a scalar value.

What are the verification errors you are getting for this? This should pass verification for `i32`. If it fails for smaller data types, that is correct, since the minimal transfer size is 4 bytes.

> Special Case: Does not accept sizes of form [1, 1, 1, N] where N >= 1024; this is a special case for npu_dma_memcpy_nd so hopefully not too hard to port

This may warrant some discussion. Making this fail was a deliberate choice on my side. I am against handling this as a special case. 

Here's my reasoning: If we were to support `[1, 1, 1, 1025]`, like `dma_memcpy_nd` currently does, this gives the appearance that this configures a data layout transformation and that e.g. `[2, 3, 4, 1025]` would also be possible. However such a transformation is not in fact representable due to hardware limits. Therefore, especially for a low-level dialect like ours that is supposed to mirror the hardware, an overflow in the lowest level should yield an error i.m.o.

Linear transfers of size 1024+ (i.e. without data layout transformations) are still possible in the DMA task syntax: Just leave off the dimensions and only specify the total transfer length. Or, alternatively, you can split up the dimensions, e.g. `sizes=[1,1,1,1024], strides=[1,1,1,1]` is not possible but `sizes=[1,1,2,512], strides=[1,1,512,1]` should work and represents the same (linear) transfer (unnecessarily complicated since it's a linear transfer but you get the gist). 

This keeps the mental model clear for users: If you supply dimensions, they will go into the registers that configure data layout transformations. If you supply dimensions that can't be expressed in hardware, you get an error. The current `dma_memcpy_nd` implementation violates this mental model and does not set the layout transformation registers if it detects the special `[1,1,1,...]` case.

Happy to hear other thoughts, but that's the reason I did things the way I did.

<sup>*: I'm trusting you on the 1024 number here as I don't know it off the top of my head. My argument still should apply even if the actual number is different.</sup>

> Functionality: It looks like all examples with a repeat count/step in the upper dimension fail; there appears to be a repeat_count parameter in the call to dma_configure_task_for which I believe works; however, this is not sufficient for examples that also have an iteration_step value in the upper dimension of the strides.

[Here](https://github.com/Xilinx/mlir-aie/blob/9519df5c492f3c8cb2b2dccce20800bbbfcd8b94/programming_examples/basic/matrix_multiplication/single_core/aie2_alt.py#L307) is an example of repeat count. It worked for me for the matmul, but it is entirely possible that it's buggy. The iteration step would be defined by the highest-dim stride [a couple lines down](https://github.com/Xilinx/mlir-aie/blob/9519df5c492f3c8cb2b2dccce20800bbbfcd8b94/programming_examples/basic/matrix_multiplication/single_core/aie2_alt.py#L317).

The highest dimension of the data layout transformations should dictate the `iteration_step` that you mentioned. Note however that the repeat count is 0 (zero) if a BD is to be executed exactly once, and 1 (one) if it is to be executed twice (i.e., repeated once), etc. The rationale why we're not using the highest dimension size as the repeat count is BD chaining: You can specify a chain of BDs, and the repeat count will apply to the *entire chain*, but I believe each BD within it can have a different iteration step (not 100% sure on this). This wasn't a consideration for `dma_memcpy_nd` because chaining is not a thing there.

**Review by andrej — COMMENTED** [review]
Looks good to me overall. I left some nitpicky comments.

**Comment by jgmelber** [comment]
> Linear transfers of size 1024+ (i.e. without data layout transformations) are still possible in the DMA task syntax: Just leave off the dimensions and only specify the total transfer length. 

@andrej I just pushed some commits to fix this issue. I believe the verifier wasn't allowing the special case for linear transfers that the hardware supports. 

I will go through more of your detailed comments tomorrow. There are still some failing tests to fix too.

File: python/dialects/aiex.py [hunk 1]
```diff
@@ -836,3 +836,64 @@ def dma_start_bd_chain_for(symbol, args, alloc, *pyargs, **kwargs):
...
+    sizes: MixedValues | None = None,
+    strides: MixedValues | None = None,
+    transfer_len: int | None = None,
+):
+    if offset is None:
+        offset = 0
+    if sizes is None:
+        sizes = [0] * 4
+    if strides is None:
+        strides = [0] * 3 + [1]
```
[the remark this answers is not in the document]
Comment (hunhoffe) [reply python/dialects/aiex.py:853]: `len` is a special function in python, e.g.,
```python
my_list = [1, 2, 3]
len(my_list)
```

It's best not to override this. We should never use `len` for arguments in python bindings, if we can avoid it.

File: python/dialects/aiex.py [hunk 2]
```diff
@@ -836,3 +836,64 @@ def dma_start_bd_chain_for(symbol, args, alloc, *pyargs, **kwargs):
...
         T.index(), chain_sym, args, alloc_sym, *pyargs, **kwargs
     )
+
+
+def shim_dma_bd(
+    mem,
+    offset: int | None = None,
+    sizes: MixedValues | None = None,
+    strides: MixedValues | None = None,
+    transfer_len: int | None = None,
```
[the remark this answers is not in the document]
Comment (hunhoffe) [reply python/dialects/aiex.py:846]: `len` is a special function in python, e.g.,
```python
my_list = [1, 2, 3]
len(my_list)
```

It's best not to override this. We should never use `len` for arguments in python bindings, if we can avoid it.

File: python/dialects/aiex.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hunhoffe) [reply python/dialects/aiex.py:853]: If we don't know the sizes, we can't calculate the length for the user. So I believe there should be defaults at the python level, unless we move length calculation lower.

[the remark this answers is not in the document]
Comment (fifield) [reply python/dialects/aiex.py:853]: > If we don't know the sizes, we can't calculate the length for the user. So I believe there should be defaults at the python level, unless we move length calculation lower.
> 
> This code was also copy/pasted from the `npu_dma_memcpy_nd` python wrapper. I think it's important for users to have consistent behavior between interfaces that do similar things, if possible.

I won't pretend I've ingested all the details here but consistent interfaces are important. One way to get that is to put things at the C++ level so that iron, normal mlir-aie, _as well as other users_ (e.g. mlir-air) all see the same behavior.

File: python/dialects/aiex.py [hunk 3]
```diff
@@ -836,3 +836,64 @@ def dma_start_bd_chain_for(symbol, args, alloc, *pyargs, **kwargs):
...
+    if strides is None:
+        strides = [0] * 3 + [1]
+
+    if transfer_len is None:
+        if len(sizes) >= 4:
+            # For shim dma bd, highest dimension is repeat count which is not included in the length
+            transfer_len = np.prod(sizes[1:])
+        else:
+            # If does not have highest dimension, then we can take the product of all dimensions
+            transfer_len = np.prod(sizes)
```
[the remark this answers is not in the document]
Comment (hunhoffe) [reply python/dialects/aiex.py:861]: In Python, we generally don't enforce having exactly 4 dimensions. Unless you want to start checking number of dimensions and outputting a useful error message in Python (duplicated from the checks in C++ code) then we need to avoid unchecked assumptions as to the length of the sizes/strides lists/

File: python/dialects/aiex.py [hunk 4]
```diff
@@ -836,3 +836,64 @@ def dma_start_bd_chain_for(symbol, args, alloc, *pyargs, **kwargs):
...
+_orig_dma_start_task = dma_start_task
+
+
+def dma_start_task(*args: DMAConfigureTaskForOp):
+    if len(args) == 0:
+        raise ValueError(
+            "dma_start_task must receive at least one DMAConfigureTaskForOp to free"
+        )
+    for dma_task in args:
+        _orig_dma_start_task(dma_task)
```
[the remark this answers is not in the document]
Comment (hunhoffe) [reply python/dialects/aiex.py:899]: There is a tradition of not forcing a user to make a list for a single element elsewhere in the python bindings (e.g., here: https://github.com/Xilinx/mlir-aie/blob/774e0e6ae680441a6ea54638901836a8a4b5d69b/python/dialects/aie.py#L441). I believe the implementation I chose is not an uncommon way to implement this pattern in python. I've also used this syntax here: https://github.com/Xilinx/mlir-aie/blob/774e0e6ae680441a6ea54638901836a8a4b5d69b/python/dialects/aiex.py#L38

Since I hadn't made other functions plural (since they work for single elements) I did not make this function plural for consistency.

I don't think adding attributes is a problem as you can easily add keyword arguments to this function still.

I personally think the current implementation is ok, but if you feel strongly I can change the implementation and naming convention of this and other functions to be consistent in another way.

**Comment by hunhoffe** [comment]
@andrej Thanks much for the review and all the information!

File: python/dialects/aiex.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hunhoffe) [reply python/dialects/aiex.py:853]: @andrej @jgmelber are there any technical reasons you can think of why the length cannot be calculated at a lower level if the user does not specify a length?

File: programming_examples/basic/dma_transpose/aie2_alt.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+                # Stride of K in dim/wrap 2 skips an entire row to read a full column
+                in_task = dma_configure_task_for(of_in, issue_token=True)
+                with bds(in_task) as bd:
+                    with bd[0]:
+                        shim_dma_bd(
+                            A,
+                            sizes=[1, 1, K, M],
+                            strides=[1, 1, 1, K],
+                        )
+                        EndOp()
```
[the remark this answers is not in the document]
Comment (hunhoffe) [reply programming_examples/basic/dma_transpose/aie2_alt.py:63]: Can confirm though trying this in an example: `EndOp` is currently necessary. 

Error without it is:
```
error: "-":21:9: block with no terminator, has "aie.dma_bd"(%arg0) <{dimensions = #aie<bd_dim_layout_array[<size = 1, stride = 1>, <size = 1, stride = 1>, <size = 32, stride = 1>, <size = 64, stride = 32>]>, len = 2048 : i32, offset = 0 : i32}> : (memref<64x32xi32>) -> ()
 note: "-":21:9: see current operation: "aie.dma_bd"(%arg0) <{dimensions = #aie<bd_dim_layout_array[<size = 1, stride = 1>, <size = 1, stride = 1>, <size = 32, stride = 1>, <size = 64, stride = 32>]>, len = 2048 : i32, offset = 0 : i32}> : (memref<64x32xi32>) -> ()
```

However, seen in the linked test, it seems like the `EndOp` is not necessary if you call `next_bd` but I haven't fully investigated: https://github.com/Xilinx/mlir-aie/blob/main/test/python/dma_tasks.py



I think with some additional time, I could clean this up. However, the python bindings I'm working on autogenerate the dma_task code, so the user will never see this anyways. At the moment, I don't see the `EndOp` cleanup as an urgent task at this time.

[the remark this answers is not in the document]
Comment (fifield) [reply programming_examples/basic/dma_transpose/aie2_alt.py:63]: > However, seen in the linked test, it seems like the EndOp is not necessary if you call next_bd

Because aie.next_bd is also a terminator.

[the remark this answers is not in the document]
Comment (andrej) [reply programming_examples/basic/dma_transpose/aie2_alt.py:63]: next_bd is also a terminator so that makes sense. I agree that this isn't important, just was curious.

File: python/dialects/aiex.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrej) [reply python/dialects/aiex.py:853]: I believe for dma_memcpy_nd we already have the length calculation at the lower level. So no I don't think there's anything preventing us from doing it at the lower level.

File: python/dialects/aiex.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrej) [reply python/dialects/aiex.py:861]: The code I suggested should work with any number of dimensions, it does not require 4.

File: python/dialects/aiex.py [hunk 6]
```diff
@@ -836,3 +836,64 @@ def dma_start_bd_chain_for(symbol, args, alloc, *pyargs, **kwargs):
...
+    if offset is None:
+        offset = 0
+    if sizes is None:
+        sizes = [0] * 4
+    if strides is None:
+        strides = [0] * 3 + [1]
+
+    if transfer_len is None:
+        if len(sizes) >= 4:
+            # For shim dma bd, highest dimension is repeat count which is not included in the length
```
[the remark this answers is not in the document]
Comment (andrej) [reply python/dialects/aiex.py:857]: I agree that the interface could use some cleaning up. For dma_memcpy_nd it made some sense to include it with the dimensions because there only ever was one BD in a task, so it behaved consistently. However, even for that, doing it that way restricts users to only be able to use the second example from above, 1 and 3 would not be possible since `dma_memcpy_nd` forces the iteration step wrap to be equal to the repeat count. For chained BDs the simplification breaks down completely i.m.o.

**Comment by andrej** [comment]
> @andrej Thanks much for the review and all the information!

Thanks for taking them into consideration. I hope I'm not getting on your nerves with the nitpicky comments, I'm just sharing my thoughts since I was asked to review. Probably none of my comments should be blocking, even if I feel strongly about some.

File: python/dialects/aiex.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrej) [reply python/dialects/aiex.py:861]: That being said your comment maybe proves that it is not, in fact, clearer. haha!

[the remark this answers is not in the document]
Comment (hunhoffe) [reply python/dialects/aiex.py:861]: Ack! You are right, whoops! I will change this. Thanks!

**Comment by hunhoffe** [comment]
> > @andrej Thanks much for the review and all the information!
> 
> Thanks for taking them into consideration. I hope I'm not getting on your nerves with the nitpicky comments, I'm just sharing my thoughts since I was asked to review. Probably none of my comments should be blocking, even if I feel strongly about some.

Of course! I am not annoyed (sorry if I seemed short or something!) and I'm glad for the comments - it's good to have another perspective on the Python code, because I'm not sure if many people have scrutinized some aspects of my previous PRs.

**Comment by hunhoffe** [comment]
Implementation plans after discussion:
- [ ] Calculate length of `bd_op` within `bd_op` constructor or similar
- [ ] With sizes now being `int32` instead of `int16`, double check max sizes are screened appropriately, including...
- [ ] Add test case to check behavior for edge case of linear transfer with size == to int max, since max size is 2**32 - 1
- [ ] Change `bd_op` to take sizes, strides instead of dimensions; create default sizes/strides in compiler (note: default should be something like [0, 0, 0, 1] and not [1, 1, 1, 1]) and remove default creation from python
- [ ] Check to make sure `bd_op` takes maximum of 3 sizes and maximum of 4 strides
- [ ] Do similar changes for removing default for `npu_dma_memcpy_nd`
- [ ] Add python wrapper function for to single chain case, because syntax can be a little clunky here; this function could safely take 4 sizes and 4 strides.

**Review by jgmelber — APPROVED** [review]

File: programming_guide/section-2/section-2g/README.md [hunk 7]
```diff
@@ -130,12 +130,83 @@ Waiting on DMAs associated with more than one object fifo:
...
+Waiting on DMAs associated with more than one object fifo:
+```python
+dma_await_task(in_task, out_task)  
+```
+
+`dma_await_task` can only be called on a task created with `issue_token=True`. If `issue_token=False` (which is default), then `dma_free_task` should be called when the programmer knows that task if complete.
+
+#### **Best Practices for Data Movement and Synchronization with `dma_task` Operations**
+
+- **Await or Free to Reuse Buffer Descriptors**: While the exact buffer descriptor (BD) used for each operation is not visible to the user with the `dma_task` operations, there are still a finite number (maximum of `16` on a Shim Tile). Thus, it is important to use `dma_await_task` or `dma_free_task` before the number of BDs are exhausted so that they may be reused.
```
Comment (andrej) [inline programming_guide/section-2/section-2g/README.md:206]: Users may wonder when to use `free` vs. when to use `await` here. I suggest adding:

> `dma_free_task` allows the compiler to reuse the buffer descriptors of a task without synchronization. Using `dma_free_task(X)` before task `X` has completed will lead to a race condition and unpredictable behavior. Only use `dma_free_task(X)` in conjunction with some other means of synchronization. For example, you may issue `dma_free_task(X)` after a call to `dma_await_task(Y)` if you can reason that task `Y` can only complete _after_ task `X` has completed.

**Review by andrej — COMMENTED** [review]
Left some comments on the programming guide

File: programming_guide/section-2/section-2g/README.md [hunk 8]
```diff
@@ -130,12 +130,83 @@ Waiting on DMAs associated with more than one object fifo:
...
+    mem,
+    tensor_tile: TensorTile | None = None,
+    offset: int | None = None,
+    sizes: MixedValues | None = None,
+    strides: MixedValues | None = None,
+    transfer_len: int | None = None,
+    issue_token: bool = False,
+)
+```
+- **`alloc`**: This is a reference to the object FIFO or the string name of an object FIFO that records a Shim Tile and one of its DMA channels allocated for the host-side memory transfer. In order to associate the memcpy operation with an object FIFO, this metadata string needs to match the object FIFO name string.
```
[the remark this answers is not in the document]
Comment (hunhoffe) [reply programming_guide/section-2/section-2g/README.md:162]: I rephrased a little, but mostly adopted as-is!

File: programming_guide/section-2/section-2g/README.md (hunk 7 above)
Comment (hunhoffe) [reply programming_guide/section-2/section-2g/README.md:206]: I added a section with function signature and example uses for `dma_free_task` and also included your suggested explanation mostly verbatim.

**Comment by hunhoffe** [comment]
@andrej Thank you for the excellent feedback! Your suggestions have been incorporated.
